miner/collator.go

```
@@ -0,0 +1,245 @@
+func (bs *blockState) Coinbase() common.Address {
+	return bs.coinbase
+}
+
+// Basefee returns the basefee for the current block
+func (bs *blockState) BaseFee() *big.Int {
+	return bs.baseFee
+}
+
+// Signer returns the block-specific signer
+func (bs *blockState) Signer() types.Signer {
+	return bs.signer
+}
+
+func (bs *blockState) Commit() {
+	w := bs.worker
+	// We don't push the pendingLogsEvent while we are mining. The reason is that
+	// when we are mining, the worker will regenerate a mining block every 3 seconds.
+	// In order to avoid pushing the repeated pendingLog, we disable the pending log pushing.
+	if !w.isRunning() && len(bs.logs) > 0 {
+
+		// make a copy, the state caches the logs and these logs get "upgraded" from pending to mined
+		// logs by filling in the block hash when the block was mined by the local miner. This can
+		// cause a race condition if a log was "upgraded" before the PendingLogsEvent is processed.
+		cpy := make([]*types.Log, len(bs.logs))
+		for i, l := range bs.logs {
+			cpy[i] = new(types.Log)
+			*cpy[i] = *l
+		}
+		w.pendingLogsFeed.Send(cpy)
+	}
+}
+
+func (bs *blockState) AddTransactions(sequence types.Transactions) error {
+	var (
+		w      = bs.worker
+		snap   = w.current.state.Snapshot()
+		err    error
+		logs   []*types.Log
+		tcount = w.current.tcount
+	)
+	for _, tx := range sequence {
+		if w.current.gasPool.Gas() < params.TxGas {
+			log.Trace("Not enough gas for further transactions", "have", w.current.gasPool, "want", params.TxGas)
+			return core.ErrGasLimitReached
+		}
+		from, _ := types.Sender(w.current.signer, tx)
+
+		// Check whether the tx is replay protected. If we're not in the EIP155 hf
+		// phase, start ignoring the sender until we do.
+		if tx.Protected() && !w.chainConfig.IsEIP155(w.current.header.Number) {
+			log.Trace("Ignoring reply protected transaction", "hash", tx.Hash(), "eip155", w.chainConfig.EIP155Block)
+			continue
+		}
```

Same here, really. Maybe this check should be done by the collator instead of the core?

---

I was thinking this check should go in the core because otherwise a buggy collator implementation could send blocks containing non-replay-protected txs to the miner.  Seems like if this is the responsibility of the collator implementation to do this check, then all custom collators will end up re-implementing this anyways right?

---

Oh sorry, right, I confused this check with another. 
In either case, just ignoring it here is not right, because then we break the guarantee of all-or-none. We instead need to return an error to the collator.